gh-105539: Explict resource management for connection objects in sqlite3 tests

erlend-aasland opened the pull request:
<!-- gh-issue-number: gh-105539 -->
* Issue: gh-105539
<!-- /gh-issue-number -->

erlend-aasland committed:
gh-105539: Explict resource management for connection objects in sqlite3 tests

vstinner commented on Lib/test/test_sqlite3/test_dbapi.py:
```diff
@@ -1659,97 +1668,87 @@ def test_cursor_executescript_too_large_script(self):
-        con.close()
-        with self.assertRaises(sqlite.ProgrammingError):
-            cur = con.cursor()
+        with self.assertRaisesRegex(sqlite.ProgrammingError, self.REGEX):
```
You may move ``self.assertRaisesRegex(sqlite.ProgrammingError, self.REGEX)`` in a helper function.

erlend-aasland replied:
Good suggestions; I'll do those before landing.

vstinner reviewed (commented):
LGTM.

I would suggest moving memory_database() to a new ``utils.py`` file. For me, it's surprising that tests import other tests. See for example ``Lib/test/test_asyncio/utils.py``.

sobolevn commented on Lib/test/test_sqlite3/test_hooks.py:
```diff
@@ -203,31 +207,35 @@ def progress():
 
 
 class TraceCallbackTests(unittest.TestCase):
+    def setUp(self):
```
This is also a pattern: maybe a test class mixin?

erlend-aasland replied:
Yeah, I thought about that myself. Let's do that now that we have the opportunity.

erlend-aasland replied:
You happy with the mixin?

sobolevn commented on Lib/test/test_sqlite3/test_userfunctions.py:
```diff
@@ -405,19 +405,19 @@ def test_func_deterministic_keyword_only(self):
-        def md5sum(t):
-            return
+        with memory_database() as dest:
+            def md5sum(t):
```
Function should be outside of the context manager, I guess.

erlend-aasland replied:
Good catch!

sobolevn commented on Lib/test/test_sqlite3/test_regression.py:
```diff
@@ -295,29 +287,31 @@ def test_convert_timestamp_microsecond_padding(self):
-                          sqlite.connect, ":memory:", isolation_level=123)
+        regex = "isolation_level must be str or None"
+        with self.assertRaisesRegex(TypeError, regex):
+            with memory_database(isolation_level=123):
```
`memory_database(isolation_level=123).__enter__()`?

This way we are more specific :)

sobolevn commented on Lib/test/test_sqlite3/test_regression.py:
```diff
@@ -238,7 +231,8 @@ def test_auto_commit(self):
         be created.
         """
-        con = sqlite.connect(":memory:", isolation_level=None)
+        with memory_database(isolation_level=None) as con:
```
This does not feel complete: can we assert something here? 🤔

erlend-aasland replied:
That test has been like that for years. I'm not sure it is really helpful anymore.

sobolevn commented on Lib/test/test_sqlite3/test_regression.py:
```diff
@@ -45,28 +45,24 @@ def test_pragma_user_version(self):
-        con = sqlite.connect(":memory:", detect_types=sqlite.PARSE_COLNAMES)
-        try:
+        with memory_database(detect_types=sqlite.PARSE_COLNAMES) as con:
             cur = self.con.cursor()
```
`cur` is not closed now, I think that `con.close()` will close it, just want to double check.

erlend-aasland replied:
Yeah, the connection object implicitly closes all open cursors in its `tp_dealloc`

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Address review

- Move test utility functions to util.py
- Use memory database mixin
- Add check() helper for closed connection tests
- Address other remarks by Nikita

vstinner reviewed (approved):
LGTM.

If you want, you can replace ``from test.test_sqlite3.util import`` with ``from .util import``.

erlend-aasland committed:
neater imports

sobolevn reviewed (approved):
My first `sqlite3` PR review ✅
